File: leetcode/js/remove-duplicates-from-sorted-list.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var deleteDuplicates = function (head) {
  const dummyNode = new ListNode(null);
  let currentNode = dummyNode;

  while (head) {
    if (currentNode.val !== head.val) {
      currentNode.next = new ListNode(head.val);
      currentNode = currentNode.next;
    }

    head = head.next;
  }

  return dummyNode.next;
};

var deleteDuplicates = function (head) {
  let currentNode = head;

  while (currentNode !== null && currentNode.next !== null) {
    if (currentNode.val === currentNode.next.val) {
      currentNode.next = currentNode.next.next;
    } else {
      currentNode = currentNode.next;
    }
  }

  return head;
};

File: leetcode/rust/remove-duplicates-from-sorted-list.rs
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
//
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }
impl Solution {
    pub fn delete_duplicates(head: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        if head.is_none() {
            return head;
        }

        let mut head = head;
        let mut current_node = head.as_mut().unwrap();

        while let Some(next_node) = current_node.next.as_mut() {
            if current_node.val == next_node.val {
                current_node.next = next_node.next.take();
            } else {
                current_node = current_node.next.as_mut().unwrap();
            }
        }

        head
    }
}
